File: collectors/admissions.py
"""Admissions: the undergraduate degree/program finder and its supporting pages.

Source: you.ubc.ca. The program finder is a client-side app whose full dataset
is inlined into /programs/ as `var programs = [...]` and `var degrees = {...}`,
so we parse those literals rather than driving the search UI. Admission
requirement, cost and deadline content lives in the WP REST pages collection.

The requirements themselves are a separate exercise -- see `admissionreqs`.
What a given program demands of you depends on where you went to school, so
they are not on the program page at all: the page ships a dropdown of provinces
and countries and fetches the matching rules over AJAX. That module drives the
same endpoint and lands the answers as tables.
"""

from __future__ import annotations

import json
import re
from typing import Any

from . import admissionreqs
from .base import Collector, Http, Output, js_literal, register, wants, wp_collection

HOST = "you.ubc.ca"
PROGRAMS_URL = f"https://{HOST}/programs/"

VAR_RE = re.compile(r"\bvar\s+([A-Za-z_$][\w$]*)\s*=\s*[\[{]")

# Front-end plumbing that happens to be declared the same way as the data.
IGNORE_VARS = re.compile(
    r"(Loc|Settings|Bundle|Vars|Strings)$|recaptcha|ajax|dataLayer|analytics|shim|nonce",
    re.I,
)

TAXONOMIES = ("program-areas", "student-groups", "ubc-life-topics", "cycle-stages", "content-types")

# Written as a nested object rather than a row per program, so flattening it to
# rows the way every other literal here is flattened produces four unlabelled
# blobs. `finances` reshapes it properly; keep the original next to that.
NESTED_VARS = {"costEstimatorData"}


def campus_ids(campuses: list[dict[str, Any]]) -> set[str]:
    """The campus term ids in scope (9 = Vancouver, 10 = Okanagan)."""
    return {
        str(term.get("term_id"))
        for term in campuses
        if wants(str(term.get("name", "")).strip().lower() or None)
    }


def on_campus(programs: list[dict[str, Any]], wanted_ids: set[str]) -> list[dict[str, Any]]:
    """Programs offered on a campus in scope; one offered on both is kept."""
    if not wanted_ids:
        return programs
    kept = []
    for program in programs:
        ids = program.get("campuses")
        ids = ids if isinstance(ids, list) else ([ids] if ids else [])
        if not ids or any(str(i) in wanted_ids for i in ids):
            kept.append(program)
    return kept


def _filter_by_campus(datasets: dict[str, list[dict[str, Any]]]) -> None:
    """Drop programs not offered on the selected campus.

    Programs reference campuses by term id, so the ids are resolved against the
    `campuses` list the page ships alongside.
    """
    campuses = datasets.get("campuses") or []
    wanted_ids = campus_ids(campuses)
    if not wanted_ids:
        return

    datasets["campuses"] = [c for c in campuses if str(c.get("term_id")) in wanted_ids]
    datasets["programs"] = on_campus(datasets.get("programs") or [], wanted_ids)


def _as_rows(value: Any) -> list[dict[str, Any]]:
    """Normalise either a list or an index-keyed object into a list of rows."""
    if isinstance(value, list):
        return [row for row in value if isinstance(row, dict)]
    if isinstance(value, dict):
        return [row for row in value.values() if isinstance(row, dict)]
    return []


@register
class Admissions(Collector):
    name = "admissions"
    title = "Admissions: degree and program finder"
    description = (
        "Every undergraduate program in UBC's program finder with its degree, campus, "
        "faculty, areas of interest and links, the degree list, and the you.ubc.ca "
        "content pages covering admission requirements, costs, deadlines and how to apply."
    )
    sources = (PROGRAMS_URL, f"https://{HOST}/wp-json/wp/v2/pages", admissionreqs.ENDPOINT)

    def collect(self, http: Http, out: Output) -> None:
        html = http.get(PROGRAMS_URL).text

        datasets: dict[str, list[dict[str, Any]]] = {}
        for name in dict.fromkeys(VAR_RE.findall(html)):
            if IGNORE_VARS.search(name) or name in NESTED_VARS:
                continue
            try:
                rows = _as_rows(js_literal(html, name))
            except (KeyError, ValueError, json.JSONDecodeError):
                continue
            if rows:
                datasets[name] = rows

        if "programs" not in datasets:
            raise RuntimeError("no `programs` dataset found on the program finder page")

        _filter_by_campus(datasets)

        out.describe(
            "programs",
            grain="one undergraduate program in the program finder",
            columns={
                "id": "you.ubc.ca post id -- the key everything else here joins on",
                "post_title": "program name",
                "link": "the program page",
                "summary": "one-paragraph description",
                "degrees": "degree term ids; resolve against degrees.json",
                "campuses": "campus term ids (9 = Vancouver, 10 = Okanagan)",
                "interests": "interest term ids; resolve against interests.json",
                "topics": "topic term ids; resolve against topics.json",
                "duration": "nominal length, e.g. {amount: 4, unit: years}",
            },
            joins=[
                "id -> admissions/requirements/program_requirements.program_id",
                "id -> finances/program_cost_estimates.program_id",
                "degrees -> admissions/degrees.term_id",
                "campuses -> admissions/campuses.term_id",
            ],
        )
        for name, rows in datasets.items():
            out.table(name, rows, source=PROGRAMS_URL)

        pages = wp_collection(http, HOST, "wp/v2/pages", params={"_embed": "0"})
        out.table("pages", pages, source=f"https://{HOST}/wp-json/wp/v2/pages")

        for taxonomy in TAXONOMIES:
            try:
                terms = wp_collection(http, HOST, f"wp/v2/{taxonomy}")
            except Exception:
                continue
            if terms:
                out.table(f"taxonomies/{taxonomy}", terms,
                          source=f"https://{HOST}/wp-json/wp/v2/{taxonomy}")

        degrees = {
            str(term.get("term_id")): term.get("name")
            for term in datasets.get("degrees", ())
        }
        admissionreqs.collect(http, out, programs=datasets["programs"], degrees=degrees)

File: src/collectors/admissions.ts
/** Admissions: the undergraduate degree/program finder and its supporting pages.
 *
 * Source: you.ubc.ca. The program finder is a client-side app whose full dataset
 * is inlined into /programs/ as `var programsListData = { programs, taxonomies }`
 * (the finder moved to slug-keyed taxonomies and dropped the old per-dataset
 * `var programs = [...]` / `var degrees = {...}` literals and their WP term
 * ids), so we parse that literal rather than driving the search UI. Admission
 * requirement, cost and deadline content lives in the WP REST pages collection.
 *
 * The requirements themselves are a separate exercise -- see `admissionreqs`.
 * What a given program demands of you depends on where you went to school, so
 * they are not on the program page at all: the page ships a dropdown of provinces
 * and countries and fetches the matching rules over AJAX. That module drives the
 * same endpoint and lands the answers as tables.
 */

import type { Http, Output } from "../base.ts";
import { jsLiteral, register, wants, wpCollection } from "../base.ts";
import { collect as collectAdmissionreqs, ENDPOINT as REQS_ENDPOINT } from "./admissionreqs.ts";

export const HOST = "you.ubc.ca";
export const PROGRAMS_URL = `https://${HOST}/programs/`;

export const VAR_RE = /\bvar\s+([A-Za-z_$][\w$]*)\s*=\s*[\[{]/g;

// Front-end plumbing that happens to be declared the same way as the data.
export const IGNORE_VARS = /(Loc|Settings|Bundle|Vars|Strings)$|recaptcha|ajax|dataLayer|analytics|shim|nonce/i;

export const TAXONOMIES = ["program-areas", "student-groups", "ubc-life-topics", "cycle-stages", "content-types"];

// Written as a nested object rather than a row per program, so flattening it to
// rows the way every other literal here is flattened produces four unlabelled
// blobs. `finances` reshapes it properly; keep the original next to that.
export const NESTED_VARS = new Set(["costEstimatorData"]);

// The program finder used to inline one `var` per dataset with WP term ids;
// you.ubc.ca now ships a single `programsListData` keyed by taxonomy slug.
export const FINDER_VAR = "programsListData";

// finder taxonomy key -> dataset stem it lands in, matching the old shape.
export const FINDER_TAXONOMIES: Readonly<Record<string, string>> = {
  campuses: "campuses",
  degree: "degrees",
  topic_areas: "topics",
  interests: "interests",
};

// Stable term id from a slug. The finder dropped WP's ids; the joins below
// (`programs` -> `finances`/`admissionreqs`) only need the two sides to match.
export function termIdFor(slug: string): number {
  let hash = 2166136261;
  for (let i = 0; i < slug.length; i += 1) {
    hash ^= slug.charCodeAt(i);
    hash = Math.imul(hash, 16777619);
  }
  return hash >>> 0;
}

/** Reshape the finder's `programsListData` into the old per-dataset shape, or
 *  null when the page still ships the pre-rename literals. */
export function programFinderDatasets(html: string): Record<string, Array<AnyJson>> | null {
  let raw: AnyJson;
  try {
    raw = jsLiteral(html, FINDER_VAR) as AnyJson;
  } catch {
    return null;
  }

  const taxonomies = (raw["taxonomies"] ?? {}) as AnyJson;
  const idBySlug: Record<string, number> = {};
  const datasets: Record<string, Array<AnyJson>> = {};
  for (const [inKey, outName] of Object.entries(FINDER_TAXONOMIES)) {
    const terms = (taxonomies[inKey] as AnyJson[] | undefined) ?? [];
    datasets[outName] = terms.map((term) => {
      const slug = String(term["slug"] ?? "");
      const id = termIdFor(slug);
      idBySlug[`${inKey}:${slug}`] = id;
      return { term_id: id, name: term["name"] ?? slug, slug, taxonomy: inKey };
    });
  }

  const toIds = (inKey: string, values: unknown): Array<number> => {
    if (!Array.isArray(values)) return [];
    return values.flatMap((value) => {
      const slug = String(value !== null && typeof value === "object" ? ((value as AnyJson)["slug"] ?? "") : value);
      const id = idBySlug[`${inKey}:${slug}`] ?? termIdFor(slug);
      return slug && id !== 0 ? [id] : [];
    });
  };

  const programs = (raw["programs"] as AnyJson[] | undefined) ?? [];
  datasets["programs"] = programs.map((program) => ({
    id: program["id"],
    post_title: program["title"],
    link: program["link"],
    summary: program["overview"] ?? "",
    is_new: program["is_new"] ?? false,
    campuses: toIds("campuses", program["campuses"]),
    degrees: toIds("degree", program["degrees"]),
    topics: toIds("topic_areas", program["topics"]),
    interests: toIds("interests", program["interests"]),
  }));
  return datasets;
}

export function campusIds(campuses: Array<AnyJson>): Set<string> {
  /** The campus term ids in scope (9 = Vancouver, 10 = Okanagan). */
  const ids = new Set<string>();
  for (const term of campuses) {
    if (
      wants(
        String(term["name"] ?? "")
          .trim()
          .toLowerCase() || null,
      )
    ) {
      ids.add(String(term["term_id"]));
    }
  }
  return ids;
}

export function onCampus(programs: Array<AnyJson>, wantedIds: Set<string>): Array<AnyJson> {
  /** Programs offered on a campus in scope; one offered on both is kept. */
  if (wantedIds.size === 0) return programs;
  const kept: Array<AnyJson> = [];
  for (const program of programs) {
    const raw = program["campuses"];
    const ids = Array.isArray(raw) ? raw : raw != null ? [raw] : [];
    if (ids.length === 0 || ids.some((id: unknown) => wantedIds.has(String(id)))) kept.push(program);
  }
  return kept;
}

export function filterByCampus(datasets: Record<string, Array<AnyJson>>): void {
  /** Drop programs not offered on the selected campus. */
  const campuses = datasets["campuses"] ?? [];
  const wantedIds = campusIds(campuses);
  if (wantedIds.size === 0) return;

  datasets["campuses"] = campuses.filter((c) => wantedIds.has(String(c["term_id"])));
  datasets["programs"] = onCampus(datasets["programs"] ?? [], wantedIds);
}

export function asRows(value: unknown): Array<AnyJson> {
  /** Normalise either a list or an index-keyed object into a list of rows. */
  if (Array.isArray(value)) {
    return value.filter((row): row is AnyJson => row !== null && typeof row === "object" && !Array.isArray(row));
  }
  if (value !== null && typeof value === "object" && !Array.isArray(value)) {
    return Object.values(value).filter(
      (row): row is AnyJson => row !== null && typeof row === "object" && !Array.isArray(row),
    );
  }
  return [];
}

type AnyJson = Record<string, unknown>;

export const Admissions = register(
  class {
    name = "admissions";
    folder = "admissions";
    title = "Admissions: degree and program finder";
    description =
      "Every undergraduate program in UBC's program finder with its degree, campus, " +
      "faculty, areas of interest and links, the degree list, and the you.ubc.ca " +
      "content pages covering admission requirements, costs, deadlines and how to apply.";
    sources = [PROGRAMS_URL, `https://${HOST}/wp-json/wp/v2/pages`, REQS_ENDPOINT];

    async collect(http: Http, out: Output): Promise<void> {
      const html = await http.getText(PROGRAMS_URL);

      const datasets: Record<string, Array<AnyJson>> = {};

      // Current finder page ships one slug-keyed literal; the per-dataset
      // literals below are the pre-rename fallback.
      const finder = programFinderDatasets(html);
      if (finder) {
        for (const [name, rows] of Object.entries(finder)) datasets[name] = rows;
      }

      const names = [...new Set([...html.matchAll(VAR_RE)].map((m) => m[1]!))];
      for (const name of names) {
        if (name === FINDER_VAR || IGNORE_VARS.test(name) || NESTED_VARS.has(name)) continue;
        try {
          const rows = asRows(jsLiteral(html, name));
          if (rows.length > 0 && datasets[name] === undefined) datasets[name] = rows;
        } catch {
          continue;
        }
      }

      if (datasets["programs"] === undefined) {
        throw new Error("no `programs` dataset found on the program finder page");
      }

      filterByCampus(datasets);

      out.describe("programs", {
        grain: "one undergraduate program in the program finder",
        columns: {
          id: "you.ubc.ca post id -- the key everything else here joins on",
          post_title: "program name",
          link: "the program page",
          summary: "one-paragraph description",
          degrees: "degree term ids; resolve against degrees.json",
          campuses: "campus term ids; resolve against campuses.json",
          interests: "interest term ids; resolve against interests.json",
          topics: "topic term ids; resolve against topics.json",
        },
        joins: [
          "id -> admissions/requirements/program_requirements.program_id",
          "id -> finances/program_cost_estimates.program_id",
          "degrees -> admissions/degrees.term_id",
          "campuses -> admissions/campuses.term_id",
        ],
      });
      for (const [name, rows] of Object.entries(datasets)) {
        await out.table(name, rows, { source: PROGRAMS_URL });
      }

      const pages = (await wpCollection(http, HOST, "wp/v2/pages", {
        params: { _embed: "0" },
      })) as Array<AnyJson>;
      await out.table("pages", pages, { source: `https://${HOST}/wp-json/wp/v2/pages` });

      for (const taxonomy of TAXONOMIES) {
        let terms: Array<AnyJson>;
        try {
          terms = (await wpCollection(http, HOST, `wp/v2/${taxonomy}`)) as Array<AnyJson>;
        } catch {
          continue;
        }
        if (terms.length > 0) {
          await out.table(`taxonomies/${taxonomy}`, terms, {
            source: `https://${HOST}/wp-json/wp/v2/${taxonomy}`,
          });
        }
      }

      const degrees: Record<string, string> = {};
      for (const term of datasets["degrees"] ?? []) {
        degrees[String(term["term_id"])] = String(term["name"] ?? "");
      }
      await collectAdmissionreqs(http, out, {
        programs: datasets["programs"],
        degrees,
      });
    }
  },
);
